### ISSUE TITLE
Add Content-Security-Policy: referrer no-referrer; to grain iframe

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#824]
References:

* https://github.com/sandstorm-io/sandstorm/wiki/Security-non-issues
* https://w3c.github.io/webappsec/specs/referrer-policy/

I'd just do this right now but I don't think I have time to do it _and_ test it this very second.

**Comment by ocdtrekkie** [comment]
Ongoing discussion about what to do with CSRF protections, I figured I'd point this out here: When trying to package a Django app for Sandstorm, this message is coming up:

> Forbidden (403)
> 
> CSRF verification failed. Request aborted.
> 
> You are seeing this message because this HTTPS site requires a “Referer header” to be sent by your Web browser, but none was sent. This header is required for security reasons, to ensure that your browser is not being hijacked by third parties.
> 
> If you have configured your browser to disable “Referer” headers, please re-enable them, at least for this site, or for HTTPS connections, or for “same-origin” requests.
> 
> If you are using the <meta name="referrer" content="no-referrer"> tag or including the “Referrer-Policy: no-referrer” header, please remove them. The CSRF protection requires the “Referer” header to do strict referer checking. If you’re concerned about privacy, use alternatives like <a rel="noreferrer" …> for links to third-party sites.

It specifically asks for the removal of this, so I wanted to reference that this will break some possible apps. (Or, in this case, it already has.)

I am curious if post-implementation of gateway, are we already effectively torching the Referrer header already, making this issue redundant/outdated?

**Comment by zenhack** [comment]
Quoting Jacob Weisz (2020-02-11 10:56:11)

>    I am curious if post-implementation of gateway, are we already
>    effectively torching the Referrer header already, making this issue
>    redundant/outdated?

We're blocking the incoming referrer header, so the app can't see it,
but if my understanding is correct, this would prevent the user's
browser from sending the referrer header when following a link from
inside the app, so they are slightly different.

**Comment by ocdtrekkie** [comment]
@zenhack Is this addressed with the new CSP, and if not, should it be?

**Comment by zenhack** [comment]
No, the new CSP does not touch this. It looks like this is actually deprecated:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/referrer

...and the new thing is `Referrer-Policy`:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy

**Comment by ocdtrekkie** [comment]
Is it still something we should address with the newer version? Or should we be closing this?

**Comment by zenhack** [comment]
We probably should still do this.

**Comment by garrison** [comment]
According to [Sandstorm's security practices](https://docs.sandstorm.io/en/latest/using/security-practices/),

> Randomized, unguessable hostnames can help mitigate certain common security bugs in apps, such as **XSRF**, **reflected XSS**, and **clickjacking** attacks. All of these attacks involve an attacker tricking the user's browser into performing actions on another site using the user's credentials. But for any of these attacks to work, the attacker must know the address to attack. If every user gets a different hostname, and indeed the hostnames change frequently, then it is much harder to launch these kinds of attacks.

... however, with current Sandstorm, when an external link within a grain is clicked, the target site is given the random hostname in the `Referer` header.  This site could then immediately use this information to attempt to launch one of the above attacks.

Setting `Referrer-Policy: same-origin` should fix this.

**Comment by garrison** [comment]
The other alternative that might make sense here is `Referrer-Policy: no-referrer`, as Sandstorm removes the `Referer` header anyway before passing the request to the grain.
